src/modules/core/filter_audiochannels.c

```diff
@@ -23,216 +23,416 @@
+	{
+		int s = 0;
+		int32_t *po = (int32_t*) out_buffer;
+		int32_t *pi = (int32_t*) in_buffer;
+		for ( s = 0; s < samples; s++ )
 		{
-			// non-interleaved - s32 or float
-			int size_avail = mlt_audio_format_size( *format, *samples, channels_avail );
-			int32_t *p = (int32_t*) new_buffer;
-			int i = *channels / channels_avail;
-			while ( i-- )
+			for ( o = 0; o < out_channels; o++ )
 			{
-				memcpy( p, *buffer, size_avail );
-				p += size_avail / sizeof(*p);
+				float value = 0.0;
+				for ( i = 0; i < in_channels; i++ )
+					value += (float)*(pi + i) * coeffs[o][i];
+				*po++ = CLAMP( value, INT32_MIN, INT32_MAX );
 			}
-			i = *channels % channels_avail;
-			if ( i )
+			pi += in_channels;
+		}
+	}
+	else if ( format == mlt_audio_f32le )
+	{
+		int s = 0;
+		float *po = (float*) out_buffer;
+		float *pi = (float*) in_buffer;
+		for ( s = 0; s < samples; s++ )
+		{
+			for ( o = 0; o < out_channels; o++ )
 			{
-				size_avail = mlt_audio_format_size( *format, *samples, i );
-				memcpy( p, *buffer, size_avail );
+				*po = 0.0;
+				for ( i = 0; i < in_channels; i++ )
+					*po += *(pi + i) * coeffs[o][i];
+				po++;
 			}
+			pi += in_channels;
 		}
-		// Update the audio buffer now - destroys the old
-		mlt_frame_set_audio( frame, new_buffer, *format, size, mlt_pool_release );
-		*buffer = new_buffer;
 	}
-	else if ( channels_avail == 6 && *channels == 2 )
+	else if ( format == mlt_audio_u8 )
 	{
-		// Downmix 5.1 audio to stereo.
-		// Mix levels taken from ATSC A/52 assuming maximum center and surround
-		// mix levels.
-		#define MIX(front, center, surr) (front + (0.707 * center) + (0.5 * surr))
-
-		// Convert to a supported format if necessary
-		mlt_audio_format new_format = *format;
-		switch( *format )
+		int s = 0;
+		uint8_t *po = (uint8_t*) out_buffer;
+		uint8_t *pi = (uint8_t*) in_buffer;
+		for ( s = 0; s < samples; s++ )
 		{
-		default:
-			// Unknown. Try to convert to float anyway.
-			mlt_log_error( NULL, "[audiochannels] Unknown format %d\n", *format );
-		case mlt_audio_float:
-		case mlt_audio_f32le:
-			new_format = mlt_audio_float;
-			break;
-		case mlt_audio_s32le:
-		case mlt_audio_s32:
-			new_format = mlt_audio_s32;
-			break;
-		case mlt_audio_s16:
-		case mlt_audio_u8:
-			new_format = mlt_audio_s16;
-			break;
-		case mlt_audio_none:
-			new_format = mlt_audio_none;
-			break;
+			for ( o = 0; o < out_channels; o++ )
+			{
+				float value = 0.0;
+				for ( i = 0; i < in_channels; i++ )
+					value += (float)*(pi + i) * coeffs[o][i];
+				*po++ = CLAMP( value, 0, UINT8_MAX );
+			}
+			pi += in_channels;
 		}
-		if ( *format != new_format && frame->convert_audio )
-			frame->convert_audio( frame, buffer, format, new_format );
+	}
+}
 
-		// Perform the downmix. Operate on the buffer in place to avoid realloc.
-		if ( *format == mlt_audio_s16 )
+static void dup_drop_channels( void* in_buffer, void* out_buffer, int in_channels, int out_channels, int samples, int format )
```

Comment:
This function is extracted from the original implementation - it duplicates or drops channels to reach the requested number of channels. No modifications here - just relocated.